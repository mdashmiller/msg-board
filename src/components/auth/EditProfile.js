import React, { Component } from 'react'
import { connect } from 'react-redux'
import { authUpdate, userUpdate } from '../../store/actions/authActions'

class EditProfile extends Component {

	state = {
		firstName: '',
		lastName: '',
		email: '',
		error: false
	}

	// component methods

	handleSubmit = e => {
		const {
			firstName,
			lastName,
			email
		} = this.state

		if (!firstName || !lastName || !email) {
			this.setState({ error: true })
		} else {
			this.props.authUpdate(email)
			this.props.userUpdate(firstName, lastName)
			this.props.history.push('/')
		}
		e.preventDefault()
	}

	handleChange = e => 
		this.setState({
			[e.target.id]: e.target.value
		})

	handleFocus = e => {
		if (e.target.value === '') {
			this.setState({
				[e.target.id]: this.props[e.target.id]
			})
		}
	}

	render() {
		return (
			<div className="container">
				<form onSubmit={this.handleSubmit} className="white">
					<h5 className="grey-text text-darken-3">Edit Profile</h5>
					<div className="input-field">
						<label htmlFor="firstName">First Name</label>
						<input type="text" id="firstName" value={this.state.firstName}
							onFocus={this.handleFocus}
							onChange={this.handleChange}
						/>
					</div>
					<div className="input-field">
						<label htmlFor="lastName">Last Name</label>
						<input type="text" id="lastName" value={this.state.lastName}
							onFocus={this.handleFocus}
							onChange={this.handleChange}
						/>
					</div>
					<div className="input-field">
						<label htmlFor="email">Email</label>
						<input type="email" id="email" value={this.state.email}
							onFocus={this.handleFocus}
							onChange={this.handleChange}
						/>
					</div>
					<div className="input-field">
						<button className="btn purple lighten-1 z-depth-0">Update</button>
					</div>
				{ this.state.error &&
					<div className="center red-text">
						<p>Please complete all fields</p>
					</div>
				}
				</form>
			</div>
		)
	}
}

const mapStateToProps = state => {
	return {
		firstName: state.firebase.profile.firstName,
		lastName: state.firebase.profile.lastName,
		email: state.firebase.auth.email
	}
}

const mapDispatchToProps = dispatch => {
	return {
		authUpdate: email => dispatch(authUpdate(email)),
		userUpdate: (firstName, lastName) => dispatch(userUpdate(firstName, lastName))
	}
}

export default connect(mapStateToProps, mapDispatchToProps)(EditProfile)
