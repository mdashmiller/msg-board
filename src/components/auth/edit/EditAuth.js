import React, { Component } from 'react'
import { connect } from 'react-redux'
import { authUpdate } from '../../../store/actions/authActions'
import PropTypes from 'prop-types'

class EditAuth extends Component {

	state = {
		email: '',
		formError: false,
		submitClicked: false,
		submitSuccess: false,
		updateAuthError: null
	}

	// component methods

	handleChange = e =>
		this.setState({
			[e.target.id]: e.target.value
		})

	handleFocus = e => {
		if (e.target.value === '') {
			// fills in the form field with focus with the current
			// value from the logged in user if the user hasn't
			// typed anything in the field yet
			this.setState({
				[e.target.id]: this.props[e.target.id]
			})
		}

		// clears error message when user
		// focuses on a form field
		this.setState({
			formError: false,
			submitClicked: false,
			submitSuccess: false,
			updateAuthError: null
		})
	}

	handleSubmit = e => {
		const { email } = this.state

		if (!email) {
			// make sure user has filled in all 
			// the form fields
			this.setState({ formError: true })
		} else {
			// call action creator
			this.props.authUpdate(email)
			this.setState({ submitClicked: true })
		}

		e.preventDefault()
	}
		
	submitSuccess = () => {
		this.setState({
			submitClicked: false,
			submitSuccess: true
		})
		setTimeout(() => this.setState({ submitSuccess: false }), 2000)
	}

	// lifecycle hooks

	componentDidUpdate(prevProps, prevState) {
		if (this.props.updateAuthError !== prevProps.updateAuthError) {
			// if there is a new updateAuthError set it in state
			this.setState({ updateAuthError: this.props.updateAuthError })
		}

		if (!this.state.updateAuthError
			&& this.props.updateAuthError
			&& this.props.updateAuthError !== prevProps.updateAuthError) {
				// if updateAuthError in state is null and the component receives a
				// a different updateAuthError than before or a new occurance 
				// of the same updateAuthError set it in state
				this.setState({ updateAuthError: this.props.updateAuthError })
		}

		if (!this.state.updateAuthError && this.state.submitClicked) {
			// if the user has clicked submit and there is no
			// updateAuthError tell the component the
			// update was a success
			this.submitSuccess()
		}
	}

	render() {
		const {
			email,
			formError,
			submitSuccess,
			updateAuthError
		} = this.state

		return (
			<form onSubmit={this.handleSubmit} id="email" className="white">
				<h5 className="grey-text text-darken-3">Update Email</h5>
				<div className="input-field">
					<label htmlFor="email">Email</label>
					<input type="email" id="email" className="field" value={email}
						onFocus={this.handleFocus}
						onChange={this.handleChange}
					/>
				</div>
				<div className="input-field">
					<button className="btn purple lighten-1 z-depth-0">Update Email</button>
					<div className="center red-text">
						{ formError && <p>Please enter a complete email address</p> }
						{ updateAuthError && <p>{ updateAuthError.message }</p> }
					</div>
					<div className="center green-text">
						{ !updateAuthError && submitSuccess && <p>Email successfully updated!</p> }
					</div>
				</div>
			</form>
		)
	}
}

const mapStateToProps = state => {
	return {
		email: state.firebase.auth.email,
		updateAuthError: state.auth.updateAuthError
	}
}

const mapDispatchToProps = dispatch => {
	return {
		authUpdate: (firstName, lastName) => dispatch(authUpdate(firstName, lastName))
	}
}

EditAuth.propTypes = {
	email: PropTypes.string,
	updateAuthError: PropTypes.shape({
		code: PropTypes.string,
		message: PropTypes.string.isRequired
	}),
	authUpdate: PropTypes.func.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(EditAuth)
