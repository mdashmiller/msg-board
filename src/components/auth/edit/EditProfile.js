import React, { Component } from 'react'
import { connect } from 'react-redux'
import { editProfile } from '../../../store/actions/authActions'
import PropTypes from 'prop-types'

class EditProfile extends Component {

	state = {
		firstName: '',
		lastName: '',
		formError: false,
		submitClicked: false,
		submitSuccess: false,
		editProfileError: null
	}

	// component methods

	handleChange = e =>
		this.setState({
			[e.target.id]: e.target.value
		})

	handleFocus = e => {
		if (e.target.value === '') {
			// fills in the form field with focus with the current
			// value from the logged in user if the user hasn't
			// typed anything in the field yet
			this.setState({
				[e.target.id]: this.props[e.target.id]
			})
		}

		// clears error message when user
		// focuses on a form field
		this.setState({
			formError: false,
			submitClicked: false,
			submitSuccess: false,
			editProfileError: null
		})
	}

	handleSubmit = e => {
		const { firstName, lastName } = this.state

		if (!firstName || !lastName) {
			// make sure user has filled in all 
			// the form fields
			this.setState({ formError: true })
		} else {
			// call action creator
			this.props.editProfile(firstName, lastName)
			this.setState({ submitClicked: true })
		}

		e.preventDefault()
	}

	submitSuccess = () => {
		this.setState({
			submitClicked: false,
			submitSuccess: true
		})
		setTimeout(() => this.setState({ submitSuccess: false }), 2000)
	}

	// lifecycle hooks

	componentDidUpdate(prevProps, prevState) {
		if (this.props.editProfileError !== prevProps.editProfileError) {
			// if there is a new editProfileError set it in state
			this.setState({ editProfileError: this.props.editProfileError })
		}

		if (!this.state.editProfileError
			&& this.props.editProfileError
			&& this.props.editProfileError !== prevProps.editProfileError) {
				// if editProfileError in state is null and the component receives a
				// a different editProfileError than before or a new occurance 
				// of the same editProfileError set it in state
				this.setState({ editProfileError: this.props.editProfileError })
		}

		if (!this.state.editProfileError && this.state.submitClicked) {
			// if the user has clicked submit and there is no
			// editProfileError tell the component the
			// update was a success
			this.submitSuccess()
		}
	}

	render() {
		const {
			firstName,
			lastName,
			formError,
			submitSuccess,
			editProfileError 
		} = this.state

		return (
			<form onSubmit={this.handleSubmit} id="user">
				<h5 className="grey-text text-darken-3">Edit Profile</h5>
				<div className="input-field">
					<label htmlFor="firstName">First Name</label>
					<input type="text" id="firstName" className="field" value={firstName}
						onFocus={this.handleFocus}
						onChange={this.handleChange}
					/>
				</div>
				<div className="input-field">
					<label htmlFor="lastName">Last Name</label>
					<input type="text" id="lastName" className="field" value={lastName}
						onFocus={this.handleFocus}
						onChange={this.handleChange}
					/>
				</div>
				<div className="input-field">
					<button className="btn z-depth-0">Update Profile</button>
					<div className="center red-text">
						{ formError && <p>Please complete all fields</p> }
						{ editProfileError && <p>{ editProfileError.message }</p> }
					</div>
					<div className="center green-text">
						{ !editProfileError && submitSuccess && <p>Profile successfully updated!</p> }
					</div>
				</div>
			</form>
		)
	}
}

const mapStateToProps = state => {
	return {
		firstName: state.firebase.profile.firstName,
		lastName: state.firebase.profile.lastName,
		editProfileError: state.auth.editProfileError
	}
}

const mapDispatchToProps = dispatch => {
	return {
		editProfile: (firstName, lastName) => dispatch(editProfile(firstName, lastName))
	}
}

EditProfile.propTypes = {
	firstName: PropTypes.string,
	lastName: PropTypes.string,
	editProfileError: PropTypes.shape({
		code: PropTypes.string,
		message: PropTypes.string.isRequired,
	}),
	editProfile: PropTypes.func.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(EditProfile)
