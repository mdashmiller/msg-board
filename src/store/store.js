// exporting store for testing purposes
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'

import rootReducer from './reducers/rootReducer'

const store = state => createStore(rootReducer, state, applyMiddleware(thunk))

export default store
