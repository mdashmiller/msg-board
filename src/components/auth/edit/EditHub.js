import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Redirect } from 'react-router-dom'
import EditProfile from './EditProfile'
import EditAuth from './EditAuth'

class EditHub extends Component {
	
	render() {
		const { auth } = this.props

		if (!auth.uid) return <Redirect to="/signin" />
			
		return (
			<div className="container">
				<EditProfile />
				<EditAuth />
			</div>
		)
	}
}

const mapStateToProps = state => {
	return {
		auth: state.firebase.auth
	}
}

export default connect(mapStateToProps)(EditHub)
