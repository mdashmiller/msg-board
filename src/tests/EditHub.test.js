import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { BrowserRouter } from 'react-router-dom'

import store from '../store/store'
import EditHub from '../components/auth/edit/EditHub'
import EditProfile from '../components/auth/edit/EditProfile'
import EditAuth from '../components/auth/edit/EditAuth'

Enzyme.configure({ adapter: new Adapter() })

const setUp = (props = {}) => {
	return shallow(
		<BrowserRouter>
			<Provider store={store()}>
				<EditHub {...props} />
			</Provider>
		</BrowserRouter>
	)
}

const props = {
	mobileNavVisible: false,
	mobileNotesVisible: false
}

describe('<EditHub /> rendering', () => {

	let wrapper
	beforeEach(() => {
		wrapper = setUp(props)
	})

	it('should render without errors', () => {
		expect(wrapper.find(EditHub).length).toBe(1)
	})

})

describe('<EditHub /> snapshot', () => {

	it('should have a valid snapshot', () => {
		const component = renderer.create(
			<BrowserRouter>
				<Provider store={store()}>
					<EditHub {...props} />
				</Provider>
			</BrowserRouter>
		)
		let tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})	

})

describe('<EditHub /> mounting and unmounting', () => {

	it('should render without crashing', () => {
	    const div = document.createElement('div')
	    ReactDOM.render(
	    	<BrowserRouter>
	    		<Provider store={store()}>
	    			<EditHub {...props} />
	    		</Provider>
	    	</BrowserRouter>
	    , div)
	    ReactDOM.unmountComponentAtNode(div)
	})

})
