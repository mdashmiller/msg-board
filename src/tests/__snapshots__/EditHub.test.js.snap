// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditHub /> rendering has a valid snapshot 1`] = `null`;

exports[`<EditHub /> rendering should have a valid snapshot 1`] = `null`;

exports[`<EditHub /> snapshot should have a valid snapshot 1`] = `null`;
